python-spec/src/somacore/spatialdata.py

```diff
@@ -0,0 +1,763 @@
+
+        As duplicate index values are not allowed, index values already present
+        in the object are overwritten and new index values are added.
+
+        Args:
+            values: An Arrow table containing all columns, including
+                the index columns. The schema for the values must match
+                the schema for the ``DataFrame``.
+
+        Returns: ``self``, to enable method chaining.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    # Metadata operations
+
+    @property
+    @abc.abstractmethod
+    def schema(self) -> pa.Schema:
+        """The schema of the data in this dataframe.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def index_column_names(self) -> Tuple[str, ...]:
+        """The names of the index (dimension) columns.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def axis_names(self) -> Tuple[str, ...]:
+        """The names of the axes of the coordinate space the data is defined on.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def domain(self) -> Tuple[Tuple[Any, Any], ...]:
+        """The allowable range of values in each index column.
+
+        Returns: a tuple of minimum and maximum values, inclusive,
+            storable on each index column of the dataframe.
+
+        Lifecycle: experimental
+        """
+        raise NotImplementedError()
+
+
+class MultiscaleImage(  # type: ignore[misc]  # __eq__ false positive
+    base.SOMAObject,
+    Generic[_DenseND, _RootSO],
+    MutableMapping[str, _DenseND],
+    metaclass=abc.ABCMeta,
+):
+    """A multiscale image with an extendable number of levels.
+
+    The multiscale image defines the top level properties. Each level must
+    match the expected following properties:
+
+    Lifecycle: experimental
+    """
+
+    # This class is implemented as a mixin to be used with SOMA classes.
+    # For example, a SOMA implementation would look like this:
+    #
+    #     # This type-ignore comment will always be needed due to limitations
+    #     # of type annotations; it is (currently) expected.
+    #     class MultiscaleImage(  # type: ignore[type-var]
+    #         ImplBaseCollection[ImplSOMAObject],
+    #         somacore.MultiscaleImage[ImplDenseNDArray, ImpSOMAObject],
+    #     ):
+    #         ...
+
+    soma_type: Final = "SOMAMultiscaleImage"  # type: ignore[misc]
+    __slots__ = ()
+
+    # Lifecycle
+
+    @classmethod
+    @abc.abstractmethod
+    def create(
+        cls,
+        uri: str,
+        *,
+        type: pa.DataType,
+        image_type: str = "CYX",
+        reference_level_shape: Sequence[int],
+        axis_names: Sequence[str] = ("c", "x", "y"),
```

Comment:
Is there any assumed connection between the image_type and the order of the axis names?

Reply:
The way I currently implemented it, there is only a connection with the location of channel, but that is really confusing (even I'm confused!). So yes, it should match the order and I'll change the order of the default `axis_names` to match the default `image_type`.